### PR TITLE
Dataflow: Add support for implicit reads

### PR DESCRIPTION
**Pull request by aschackmull**
This adds a new overridable predicate on configurations:
```ql
  /**
   * Holds if an arbitrary number of implicit read steps of content `c` may be
   * taken at `node`.
   */
  predicate allowImplicitRead(Node node, Content c)
```
The purpose is to support sinks and taint steps accepting non-empty accesspaths.

I've also added a default override in taint-tracking configurations, which the language specific implementations can use to add such default support. This is just a (hopefully helpful) default and can be overridden by specific configurations. For Java, the default implementation is all array-, collection-, and map-value-read steps that match the type at the given sink / additional taint step.

This means that we ought to be able to drop the temporary store-as-taint workaround, which was costing quite some performance.

### CODE REVIEW

**Comment by asgerf** [comment]
Quick question about how this relates to something we do in the JS libraries. We support custom load steps via:
```codeql
  /**
   * EXPERIMENTAL. This API may change in the future.
   *
   * Holds if the property `prop` of the object `pred` should be loaded into `succ`.
   */
  predicate isAdditionalLoadStep(DataFlow::Node pred, DataFlow::Node succ, string prop)
```
I'm wondering if `allowImplicitRead(node, c)` would be equivalent to `isAdditionalLoadStep(node, node, c)`, or is there something else going on here?

**Comment by aschackmull** [comment]
> I'm wondering if `allowImplicitRead(node, c)` would be equivalent to `isAdditionalLoadStep(node, node, c)`, or is there something else going on here?

It is almost equivalent, yes. There's a slight difference if you also use type pruning, because this would easily remove the flow you gained from `isAdditionalLoadStep(node, node, c)` (the types don't match up). So conceptually you may think of this as an in-place additional read step with the quirks related to type pruning sorted such that it actually works.

**Comment by smowton** [comment]
Started a benchmark to check effects on codeql-go (without any default implementation)

**Comment by aschackmull** [comment]
Java test of 38319a483212d740c28bbfe6008a89704e9d8d90 (just the dataflow lib changes) https://jenkins.internal.semmle.com/job/Changes/job/Java-Differences/1464/

Testing PR head (c06e152e900d9d51db6df0b3a262e7cfdd2f01ab):
https://jenkins.internal.semmle.com/job/Changes/job/Java-Differences/1465/
https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/1149/
https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/2099/
https://jenkins.internal.semmle.com/job/Changes/job/Python-Differences/585/

**Review by smowton — APPROVED** [review]
Go benchmark results are good (either neutral or slightly beneficial)

**Comment by aschackmull** [comment]
Java perf measurements are thrown off by a case of bad magic. I'll need to fix it and restart them.

**Comment by aschackmull** [comment]
Looks like the performance regression by the store-as-taint workaround is indeed fixed:
```
Security/CWE/CWE-090/LdapInjection.ql	363	138	-225
Security/CWE/CWE-078/ExecUnescaped.ql	477	162	-315
Security/CWE/CWE-078/ExecTainted.ql	481	155	-326
```
But the new feature that's added to the dataflow library to do this does come with an added performance overhead - mainly due to tuple-numbering, as far as I have been able to see.

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll [hunk 1]
```diff
@@ -788,6 +788,66 @@ predicate localInstructionFlow(Instruction e1, Instruction e2) {
...
+}
+
+/** A reference through an array. */
+class ArrayContent extends Content, TArrayContent {
+  override string toString() { result = "[]" }
+}
+
+/** A reference through the contents of some collection-like container. */
+private class CollectionContent extends Content, TCollectionContent {
+  override string toString() { result = "<element>" }
```
Comment (hvitved) [inline cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll:848]: Why did you change the C++ content `toString()`s?

**Review by hvitved — COMMENTED** [review]
Looks solid to me.

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll (hunk 1 above)
Comment (aschackmull) [reply cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll:848]: To match Java. They were originally just copy-pasted from Java and weren't used in either language. When I added support for collection-flow in Java I updated them to something I thought made more sense (but the old somewhat arbitrary values still lingered in C++). So now that I moved them I thought I might as well update them. I did check with @MathiasVP beforehand that this was ok with him.

**Review by hvitved — APPROVED** [review]

**Comment by MathiasVP** [comment]
> https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/2106/

CPP-Differences doesn't look super good. There's a clear performance regression on all of the dataflow queries. Is such a big slowdown expected?

Screenshot 2021-06-24 at 14 39 18

**Comment by aschackmull** [comment]
> Is such a big slowdown expected?

I'd say unfortunately yes for the time being. It seems like all dataflow queries are taking a hit across all languages. I've been trying to dig into it, but there's no clear single culprit - rather it seems like the wrapping of `Node` as `TNodeNormal` with all the necessary accompanying joins are to blame: It's a bunch of additional `Node`-sized scans and joins and tuplenumberings. On the plus side, I _think_ a lot of the additional overhead can be run in parallel, and I'm not sure whether the LANG-Differences jobs does that (so the real cost may be less severe).

I do have some ideas for improving the situation, but I'd rather look into those as follow-up work.

**Review by MathiasVP — APPROVED** [review]
Thanks for the explanation!
